lib/multitimer/src/lib.rs

```diff
@@ -0,0 +1,407 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! A timer multiplexer.
+//!
+//! `Multitimer` lets you wrap a single underlying event timer and treat it as
+//! multiple independent event timers. The independent event timers correspond
+//! to variants of an enum type, to make it easy to tell them apart.
+//!
+//! The expected usage model is:
+//!
+//! - Create an `enum` type naming your timers, and derive the `Enum` trait
+//!   (from the `enum_map` crate) for it.
+//!
+//! - Create a `Multitimer<YourEnumType>`.
+//!
+//! - Use its API to configure your timers to your heart's content.
+//!
+//! - When notifications arrive from the underlying timer, feed them into
+//!   `Multitimer::handle_notification`.
+//!
+//! - When you're ready to process timer events (which may or may not be
+//!   immediately after the notification), call `Multitimer::iter_fired`.
+//!
+//! **Note:** the `Multitimer` assumes that it has sole control of the
+//! underlying timer. If you create two `Multitimer`s using the same underlying
+//! timer, they will fight and the results will be unpleasant. API like
+//! `sleep_until`/`sleep_for` that saves and restores timer settings _can_ be
+//! used alongside `Multitimer`.
+
+#![cfg_attr(target_os = "none", no_std)]
+
+use enum_map::{EnumArray, EnumMap};
+
+// Import the actual syscalls if we're targeting actual Hubris; otherwise we use
+// some stub functions defined below.
+#[cfg(target_os = "none")]
+use userlib::{sys_get_timer, sys_set_timer};
+
+pub struct Multitimer<E: EnumArray<Timer>> {
+    notification_bit: u8,
+    current_setting: Option<u64>,
+    timers: EnumMap<E, Timer>,
+}
+
+impl<E: EnumArray<Timer> + Copy> Multitimer<E> {
+    pub fn new(notification_bit: u8) -> Self {
+        Self {
+            notification_bit,
+            current_setting: None,
+            timers: EnumMap::default(),
+        }
+    }
+
+    /// Sets the timer chosen by `which` to go off at time `deadline`, with
+    /// optional auto-repeat behavior. This replaces any prior setting for the
+    /// timer and enables it.
+    ///
+    /// This operation may cause a syscall: if `deadline` is sooner than all the
+    /// other deadlines being managed by this multitimer, we will call into the
+    /// OS to move our timer forward.
+    pub fn set_timer(
+        &mut self,
+        which: E,
+        deadline: u64,
+        repeat: Option<Repeat>,
+    ) {
+        // If the timer has previously fired without us noticing it, preserve
+        // that across set.
+        let fired_but_not_observed = self.timers[which].fired_but_not_observed;
+        self.timers[which] = Timer {
```

Comment:
Vague question: is it better to save `fired_but_not_observed` then build a whole new `Timer`, vs modifying the deadline on the existing timer, e.g.
```rust
self.timers[which].deadline = Some((deadline, repeat))
```

Reply:
I could go either way -- my current approach feels more robust if Timer grows new fields in the future, while your approach is simpler with the current definition.